aria/workflow_engine/engine.py

```diff
@@ -68,57 +68,83 @@ def create_workflow(
             workflow_context,
             graph,
             log_indentation_count=0,
-            start_task=StartWorkflowTask,
-            end_task=EndWorkflowTask,
-            dependent_tasks=()):
-        # TODO: refactor the function to a more readable function
+            start_task_cls=StartWorkflowTask,
+            end_task_cls=EndWorkflowTask,
+            depends_on=()):
+        # TODO: refactor the function to a more readable one
         assert isinstance(graph, TaskGraph)
 
-        self.workflow_tasks['{0}.start'.format(graph.name)] = type(
-            '{0}.start'.format(graph.name),
-            (start_task,),
-            {})(context=workflow_context)
-
-        for operation_or_sub_workflow in graph.tasks_in_order(reverse=True):
-            if self._task_is_operation(operation_or_sub_workflow):
-                self._create_operation_task(graph, operation_or_sub_workflow)
-                continue
-            sub_workflow_graph = operation_or_sub_workflow
-            self.create_workflow(
-                log_indentation_count=log_indentation_count + 1,
-                graph=sub_workflow_graph,
-                workflow_context=workflow_context,
-                start_task=StartSubWorkflowTask,
-                end_task=EndSubWorkflowTask,
-                dependent_tasks=self._operation_requires(graph, sub_workflow_graph))
-
-        workflow_requires = self._workflow_requires(graph.tasks)
-        workflow_requires.extend(dependent_tasks)
-
-        self.workflow_tasks[graph.name] = type(
-            graph.name,
-            (end_task,),
-            {})(context=workflow_context, depends_on=workflow_requires)
-
-    def _create_operation_task(self, graph, operation_context):
+        workflow_start_task = self.workflow_tasks[self._start_task_name(graph.name)] = type(
+            self._start_task_name(graph.name),
+            (start_task_cls,),
+            {})(context=workflow_context, depends_on=depends_on)
+
+        for operation_or_workflow, dependencies in graph.tasks_in_order(
+                reverse=True, dependencies=True):
```

Comment:
for operation_or_workflow, dependencies in graph.tasks_tree(reverse=True):
    ...